### ISSUE TITLE
Resize operator need to support 4-D inputs with more than 1 batch and 1 channel

### ISSUE THREAD

**Issue by yangia** [microsoft/onnxruntime#2463]
**Is your feature request related to a problem? Please describe.**
I use the `interpolate` operator in my pytorch model to upsample a tensor, named `x`. `x` is a 4-D tensor, NCHW format. The followings are my tirals.
1. I converted the pytorch model `xxx.pth` into `xxx.onnx`.  `xxx.onnx` was checked successfully. But inference `xxx.onnx` with onnxruntime was failed. 
`onnxruntime/core/providers/cpu/tensor/upsample.h:221 void onnxruntime::UpsampleBase::ScalesValidation(const std::vector<float>&, onnxruntime::UpsampleMode) const scales.size() == 2 || (scales.size() == 4 && scales[0] == 1 && scales[1] == 1) was false. 'Linear' mode and 'Cubic' mode only support 2-D inputs ('Bilinear', 'Bicubic') or 4-D inputs with the corresponding outermost 2 scale values being 1 in the Resize operator`
2. I sliced my `x` to be NxC `y`s, with `y` to be a `(1,1,H,W)` tensor. Then I `interpolate` `y`s to get `Y`s, and concate `Y`s to get `X`. Finally I converted the new model and the check and the inference are successful. The outputs of the converted model were also correct. However, the newly converted `xxx_new.onnx` was overly complex with hundreds of Resize operators.
So, I highly recommand that Resize operator should support 4-D inputs with more than 1 batch and 1 channel.

**System information**
- ONNX Runtime version (you are using): built form source, 21/11/2019

**Describe the solution you'd like**
Resize operator supports 4-D inputs with more than 1 batch and 1 channel

**Describe alternatives you've considered**
no alternative

**Comment by hariharans29** [comment]
AFAIK - we already support 4-D tensor in NCHW format resizing where batch > 1 and scale > 1. The only limitation is imposed on **scales** (not the input itself). For such a 4-D input, outermost scales must be both 1 (i.e.) we support "bilinear" resizing (outermost scales requiring to be 1) for multi-channel  multi-batch inputs.

To simplify:

1) Input shape that is currently accepted by ORT:
input_shape = [2, 3, 224, 224] (2 images each with 3 channels)
scales_shape = [1, 1, x, x]

Looks like this is what you need - and this is supported

2) Input shape that is valid but not yet supported by ORT:
input_shape = [2, 3, 224, 224] (2 images each with 3 channels)
scales_shape = [2, 2, x, x]

This is not supported yet as "quadrilinear" resizing is not supported yet.

Sharing a simple single node model (with just Resize node) and some toy expected inputs/ expected outputs will help.

**Comment by yangia** [comment]
> AFAIK - we already support 4-D tensor in NCHW format resizing where batch > 1 and scale > 1. The only limitation is imposed on **scales** (not the input itself). For such a 4-D input, outermost scales must be both 1 (i.e.) we support "bilinear" resizing (outermost scales requiring to be 1) for multi-channel multi-batch inputs.
> 
> To simplify:
> 
> 1. Input shape that is currently accepted by ORT:
>    input_shape = [2, 3, 224, 224] (2 images each with 3 channels)
>    scales_shape = [1, 1, x, x]
> 
> Is this what you need ?
> 
> 1. Input shape (that is valid per spec) but not yet supported by ORT:
>    input_shape = [2, 3, 224, 224] (2 images each with 3 channels)
>    scales_shape = [2, 2, x, x]
> 
> This is not supported yet as "quadrilinear" resizing is not supported yet.
> 
> Sharing a simple single node model (with just Resize node) and some toy expected inputs/ expected outputs will help.

The Resize operator was conveted from pytorch `interpolate` operator. So I cannot verify if your suggestion 1 works. What I need is definitely your suggestion 1.

**Comment by yangia** [comment]
I found the convertion script in pytorch symbolic_opset11.py as this
```
def _interpolate(name, dim, interpolate_mode):
    def symbolic_fn(g, input, output_size, align_corners=None):
        align_corners = sym_help._maybe_get_scalar(align_corners)
        output_size = sym_help._maybe_get_const(output_size, 'is')
        if sym_help._is_value(output_size):
            offsets = g.op("Constant", value_t=torch.ones(2, dtype=torch.int64))
            output_size = g.op("Cast", output_size, to_i=sym_help.cast_pytorch_to_onnx["Long"])
            output_size = g.op("Concat", offsets, output_size, axis_i=0)
        else:
            output_size = [1 if i < 2 else output_size[-(dim - i)] for i in range(0, dim)]
            output_size = g.op("Constant", value_t=torch.tensor(output_size))
        coordinate_transformation_mode = "asymmetric" if interpolate_mode == "nearest" \
            else "align_corners" if align_corners else "pytorch_half_pixel"
        empty_tensor = g.op("Constant", value_t=torch.tensor([], dtype=torch.float32))
        return g.op("Resize",
                    input,
                    empty_tensor,  # roi only takes effect whith coordinate_transformation_mode="tf_crop_and_resize"
                    empty_tensor,  # scales is not needed since we are sending out_size
                    output_size,
                    coordinate_transformation_mode_s=coordinate_transformation_mode,
                    cubic_coeff_a_f=-0.75,  # only valid when mode="cubic"
                    mode_s=interpolate_mode,  # nearest, linear, or cubic
                    nearest_mode_s="floor")  # only valid when mode="nearest"
    return symbolic_fn
```
Here we can notice that `scales` are not given in the conversion. So the question is, why ORT still check the `scales`? and where the `scales` under checkment comes from? I looked into the `scales` and found that it is 4-D, and I have no idea where it comes from.

**Comment by hariharans29** [comment]
Please check the ONNX spec for Resize - https://github.com/onnx/onnx/blob/master/docs/Operators.md#Resize. It has an input 'scales' that defines the final output shape of the resized tensor.

Any conversion framework (in this case - the PyTorch exporter) if generating an ONNX model, should comply with the ONNX spec and hence the final model will have scales in it, even if you can't see it in the conversion snippet you pasted.  (I think scales is being computed by the given `output_size` and the `input` in `g.op()`). I

If it is case-1 you need, it is supported by ORT. Please share the converted ONNX model you get from PyTorch and the expected input and output and someone will help take a look. Thanks.

**Comment by longtaohuang** [comment]
@yangia I have met the same problem as you described, and I tried to solve it with your trials, in my solution I used torch.split and torch.cat, the solution to traversal the sliced tensor I used is FOR structure.
     Then I got the error: TracerWarning: Converting a tensor to a Python index might cause the trace to be incorrect. We can't record the data flow of Python values, so this value will be treated as a constant in the future. This means that the trace might not generalize to other inputs!
    Have you met this problem and how did you solve it? Thanks so much if you can reply me.

**Comment by emilianavt** [comment]
I worked around this issue by modifying `torch/onnx/symbolic_opset11.py` as follows.

Original code in `_interpolate`:

            empty_tensor = g.op("Constant", value_t=torch.tensor([], dtype=torch.float32))
            return g.op("Resize",
                        input,
                        empty_tensor,  # roi only takes effect whith coordinate_transformation_mode="tf_crop_and_resize"
                        empty_tensor,  # scales is not needed since we are sending out_size
                        output_size,
                        coordinate_transformation_mode_s=coordinate_transformation_mode,
                        cubic_coeff_a_f=-0.75,  # only valid when mode="cubic"
                        mode_s=interpolate_mode,  # nearest, linear, or cubic
                        nearest_mode_s="floor")  # only valid when mode="nearest"

Modified code:

            empty_tensor = g.op("Constant", value_t=torch.tensor([], dtype=torch.float32))
            scales = sym_help._interpolate_size_to_scales(g, input, output_size, dim)
            return g.op("Resize",
                        input,
                        empty_tensor,  # roi only takes effect whith coordinate_transformation_mode="tf_crop_and_resize"
                        scales,  # scales is not needed since we are sending out_size
                        coordinate_transformation_mode_s=coordinate_transformation_mode,
                        cubic_coeff_a_f=-0.75,  # only valid when mode="cubic"
                        mode_s=interpolate_mode,  # nearest, linear, or cubic
                        nearest_mode_s="floor")  # only valid when mode="nearest"

However, I believe this is actually a bug in onnxruntime. The ONNX specs say:

> sizes (optional) : tensor(int64)
> The size of the output tensor. The number of elements of 'sizes' should be the same as the rank of input 'X'. **Only one of 'scales' and 'sizes' can be specified.**

This implies that requiring scales or both scales and sizes as onnxruntime does is not correct.

**Comment by yangia** [comment]
> @yangia I have met the same problem as you described, and I tried to solve it with your trials, in my solution I used torch.split and torch.cat, the solution to traversal the sliced tensor I used is FOR structure.
> Then I got the error: TracerWarning: Converting a tensor to a Python index might cause the trace to be incorrect. We can't record the data flow of Python values, so this value will be treated as a constant in the future. This means that the trace might not generalize to other inputs!
> Have you met this problem and how did you solve it? Thanks so much if you can reply me.

I didn't use torch.split. I silced with `for`, and then concatenated with torch.cat. The conversion of torch.split seems to have something wrong with it. Not sure.

**Comment by yangia** [comment]
> Due to slow CPU inference speed of Pytorch on Windows, I decided to switch to doing inference with onnxruntime, which is much faster. However, I also encountered this issue while converting my model. I worked around it by modifying `torch/onnx/symbolic_opset11.py` as follows.
> 
> Original code in `_interpolate`:
> 
> ```
>         empty_tensor = g.op("Constant", value_t=torch.tensor([], dtype=torch.float32))
>         return g.op("Resize",
>                     input,
>                     empty_tensor,  # roi only takes effect whith coordinate_transformation_mode="tf_crop_and_resize"
>                     empty_tensor,  # scales is not needed since we are sending out_size
>                     output_size,
>                     coordinate_transformation_mode_s=coordinate_transformation_mode,
>                     cubic_coeff_a_f=-0.75,  # only valid when mode="cubic"
>                     mode_s=interpolate_mode,  # nearest, linear, or cubic
>                     nearest_mode_s="floor")  # only valid when mode="nearest"
> ```
> 
> Modified code:
> 
> ```
>         empty_tensor = g.op("Constant", value_t=torch.tensor([], dtype=torch.float32))
>         scales = sym_help._interpolate_size_to_scales(g, input, output_size, dim)
>         return g.op("Resize",
>                     input,
>                     empty_tensor,  # roi only takes effect whith coordinate_transformation_mode="tf_crop_and_resize"
>                     scales,
>                     coordinate_transformation_mode_s=coordinate_transformation_mode,
>                     cubic_coeff_a_f=-0.75,  # only valid when mode="cubic"
>                     mode_s=interpolate_mode,  # nearest, linear, or cubic
>                     nearest_mode_s="floor")  # only valid when mode="nearest"
> ```
> 
> However, I believe this is actually a bug in onnxruntime. The ONNX specs say:
> 
> > sizes (optional) : tensor(int64)
> > The size of the output tensor. The number of elements of 'sizes' should be the same as the rank of input 'X'. **Only one of 'scales' and 'sizes' can be specified.**
> 
> This implies that requiring scales or both scales and sizes as onnxruntime does is not correct.

Thank you for your reply. I will try it later.

**Comment by TyrionChou** [comment]
@yangia hi, may i see your change code that use for and cat() to solve this question？
